### DISCUSSION TITLE
How to create a list with interpolated variables?

### DISCUSSION THREAD

**Discussion by mjreed-wbd** [gilch/hissp#102] [Q&A] [answered]
What's the Hissp equivalent of this Scheme?

      (define foo 'bar)
      (list 1 foo 2)

In Python it's just `(1, foo, 2)`... but I don't know how to get that out of Hissp's reader.

**Comment by gilch** [comment]
[Hidden because this comment is very out of date.]
***
**TL;DR**: install master from github, use `(hissp.basic.._macro_.prelude)` at the top of your file, then use `entuple` instead of `list`.
***
`(list ...)` will call Python's `list` builtin, which takes a single iterable argument. This results in a Python list, which is not a linked list and is not the same type as a tuple. Lists of either sort do not represent invocations in Hissp. Hissp uses tuples for that.

## Python Injection

If you know how to do something with Python expressions, you can use the inject macro to do it with Python syntax:
```python
#> .#"(1, foo, 2)"
>>> (1, foo, 2)
(1, 'bar', 2)
````
I usually don't recommend using injections when you don't have to, because it has certain pitfalls, but it's a nice escape hatch for those familiar with Python, and does have legitimate uses if you know what you are doing.

## Template

You can interpolate values into tuples with the template quote syntax, which is similar to Scheme's quasiquote:
```racket
(define foo 'bar)
```
```python
#> `(1 ,foo 2)
>>> (lambda * _: _)(
...   (1),
...   foo,
...   (2))
(1, 'bar', 2)
```
The primary use for templates is for metaprogramming and it has some features that may get in the way if you're trying to use it for data. You may want to make a helper function for that kind of thing instead. You can see from the compiled Python how templates do it: `lambda *_: _`. You can make a helper function the same way.

## Helper Function

This function was an example in the [tutorial](https://hissp.readthedocs.io/en/latest/tutorial.html#collection-atoms). It will probably work the way you expected `list` to.

```racket
(define entuple
  (lambda (: :* xs)
    xs))
```
```python
#> (entuple 1 foo 2)
>>> entuple(
...   (1),
...   foo,
...   (2))
(1, 'bar', 2)
```
As of #113, the basic prelude will define this function for you, as well as a family of related functions called the en- group.
Use `(hissp.basic.._macro_.prelude)` at the top of your .lissp files to get the prelude. This also works in the REPL, but since it starts with the basic macros loaded, `(prelude)`, without the qualification, will suffice. (The prelude is also implied for commands run with `$ lissp -c`.)

As of #140, [`hissp.basic.._macro_.en#`](https://hissp.readthedocs.io/en/latest/hissp.basic.html#hissp.basic._macro_.enQzHASH_) is a generalization of the en- group to any function of one iterable, a common occurrence in Python.
```
#> (en#tuple 1 foo 2)
>>> (lambda *_xs_QzNo4_:
...   tuple(
...     _xs_QzNo4_))(
...   (1),
...   foo,
...   (2))
(1, 'bar', 2)
```

## Pyrsistent

Another option is to create an iterable some other way, and then covert it to a tuple using Python's `tuple` builtin.

Pyrsistent is one of the libraries I recommend using with Hissp. It has helper functions similar to `entuple`, which collect multiple arguments into an iterable data structure. Given that, you can do
```python
#> (pyrsistent..l 1 foo 2) ; Pyrsistent linked-list type.
>>> __import__('pyrsistent').l(
...   (1),
...   foo,
...   (2))
plist([1, 'bar', 2])
#> (tuple _) ; Like Python, in the REPL, _ is the last value that wasn't None.
>>> tuple(
...   _)
(1, 'bar', 2)
```
or
```python
#> (tuple (pyrsistent..v 1 foo 2)) ; Through Pyrsistent's vector type.
>>> tuple(
...   __import__('pyrsistent').v(
...     (1),
...     foo,
...     (2)))
(1, 'bar', 2)
```

**Comment by gilch** [comment, marked as answer]
Nearest equivalent (this needs the bundled macros):
```racket
(define foo 'bar)
(en#tuple 1 foo 2)
```
Hissp code is based on tuples, not linked lists. Python's builtin `tuple` takes a single iterable argument. The `en#` reader macro uses a lambda to convert a function applicable to one tuple to a function of its elements.
***
Without any bundled macros (Scheme calls this "quasiquoting"):
```racket
(.update (globals) : foo 'bar)
`(,1 ,foo ,2)
```
This uses Python builtins to accomplish the assignment and the template quote reader syntax to make a data tuple. Notice how this also expands to a lambda. (Try it in the REPL.) The ints technically don't have to be unquoted here:
```racket
`(1 ,foo 2)
```
But make sure you understand [how fragment quoting works](https://github.com/gilch/hissp/wiki/Lissp-Syntax-Quick-Reference#fragments) before omitting unquotes on strings.
***
With the [empty fragment hack](https://github.com/gilch/hissp/wiki/Lissp-tricks-and-idioms#pseudo-special-forms) (`(||` `||)` brackets, no macros):
```racket
(.update (globals) : foo 'bar)
(|| 1 foo 2 ||)
```
No reader syntax either. An equivalent form also works in readerless mode:
```python
('.update',('globals',),':','foo',('quote','bar'),)
('',1,'foo',2,'',)
```
***
With the "list of" macro `@`:
```racket
(define foo 'bar)
(@ 1 foo 2)
```
This may be more idiomatic in Lissp, depending on what you are trying to do. It makes a Python list, not a tuple, however.
***
With Python fragments (escape hatch):
```python
|(foo := 'bar')|
|(1, foo, 2)|
```
Lissp allows you to inject Python expressions (or even statements, at the top level) directly into the compiled output. Anything you know how to do in Python can be done in Lissp this way.
